src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,276 @@
+#include <algorithm>
+#include <chrono>
+#include <unordered_set>
+
+#include <base/getFQDNOrHostName.h>
+#include <base/logger_useful.h>
+
+#include <Common/DNSResolver.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/setThreadName.h>
+
+#include <Core/ServerUUID.h>
+
+#include <Interpreters/Cluster.h>
+#include <Interpreters/ClusterDiscovery.h>
+#include <Interpreters/Context.h>
+
+namespace DB
+{
+
+namespace
+{
+
+constexpr size_t MAX_QUEUE_SIZE = 16;
+constexpr UInt64 QUEUE_OP_TIMEOUT_MS = 1000;
+
+fs::path getReplicasListPath(const String & zk_root)
+{
+    return fs::path(zk_root + "/replicas");
+}
+
+}
+
+ClusterDiscovery::ClusterDiscovery(
+    const Poco::Util::AbstractConfiguration & config,
+    ContextMutablePtr context_,
+    const String & config_prefix)
+    : context(context_)
+    , node_name(toString(ServerUUID::get()))
+    , server_port(context->getTCPPort())
+    , queue(std::make_shared<UpdateQueue>(MAX_QUEUE_SIZE))
+    , log(&Poco::Logger::get("ClusterDiscovery"))
+{
+    Poco::Util::AbstractConfiguration::Keys config_keys;
+    config.keys(config_prefix, config_keys);
+
+    for (const auto & key : config_keys)
+    {
+        String path = config.getString(config_prefix + "." + key + ".path");
+        trimRight(path, '/');
+        clusters_info.emplace(key, ClusterInfo(key, path));
+    }
+}
+
+/// List node in zookeper for cluster
+Strings ClusterDiscovery::getNodeNames(zkutil::ZooKeeperPtr & zk,
+                                       const String & zk_root,
+                                       const String & cluster_name,
+                                       int * version,
+                                       bool set_callback)
+{
+    auto watch_callback = [cluster_name, queue=queue, log=log](const Coordination::WatchResponse &)
+    {
+        if (!queue->tryPush(cluster_name, QUEUE_OP_TIMEOUT_MS))
+        {
+            if (queue->isFinished())
+                return;
+            LOG_WARNING(log, "Cannot push update request for cluster '{}'", cluster_name);
+        }
+    };
+
+    Coordination::Stat stat;
+    Strings nodes = zk->getChildrenWatch(getReplicasListPath(zk_root), &stat, set_callback ? watch_callback : Coordination::WatchCallback{});
+    if (version)
+        *version = stat.cversion;
+    return nodes;
+}
+
+/// Reads node information from scpecified zookeper nodes
+ClusterDiscovery::NodesInfo ClusterDiscovery::getNodes(zkutil::ZooKeeperPtr & zk, const String & zk_root, const Strings & nodes)
+{
+    NodesInfo result;
+    for (const auto & node_uuid : nodes)
+    {
+        bool ok = zk->tryGet(getReplicasListPath(zk_root) / node_uuid, result[node_uuid]);
+        if (!ok)
+        {
+            result.erase(node_uuid);
+            LOG_WARNING(log, "Cluster configuration was changed during update, skip nonexisting node");
```

Comment:
Session may expire somewhere here. Message will be misleading.
Need to check other places for that.

Reply:
Changed logic a bit, now I just return empty here and it indicates some error on update, then it won't marked as updated and will retried